spec.bs

```diff
@@ -1226,6 +1254,280 @@ On the other hand, methods for getting data from the [=shared storage database=]
+
+    1. Set |window| to |request|’s [=request/window=].
+    1. If |window| is not an [=environment settings object=] whose [=global object=] is a {{Window}}, return false.
+    1. If the result of running [=determine whether shared storage is allowed=] for |window| is false, return false.
+    1. Let |document| be |window|'s [=global object=]'s [=associated document=].
+    1. Return the result of running [=Is feature enabled in document for origin?=] on "[=PermissionsPolicy/shared-storage=]", |document|, and |request|'s [=request/current URL=]'s [=url/origin=].
+
+    Issue: The [=determine whether a request can currently use shared storage=] algorithm needs to take into account "opt-in features", as articulated in <a href="https://github.com/w3c/webappsec-permissions-policy/pull/499">https://github.com/w3c/webappsec-permissions-policy/pull/499</a>.
+  </div>
+
+
+  <div algorithm>
+    To <dfn>append or modify a Shared-Storage-Writable request header</dfn>, given a [=/request=] |request|, perform the following steps:
+
+    1. If |request|'s [=request/shared storage writable=] is not true, then return.
+    1. If the result of running [=determine whether a request can currently use shared storage=] on |request| is false, [=header list/delete=] "[=Shared-Storage-Writable=]" from |request|'s [=request/header list=].
+    1. Otherwise, [=header list/set=] ("[=Shared-Storage-Writable=]", "?1") in |request|'s [=request/header list=].
+  </div>
+
+  <div algorithm>
+    To <dfn>handle a Shared-Storage-Write response</dfn>, given a [=/response=] |response| and a [=/request=] |request|, perform the following steps:
+
+    1. If |request|'s [=request/header list=] does not [=list/contain=] "[=Shared-Storage-Writable=]", then return.
+    1. Set |window| to |request|’s [=request/window=].
+    1. [=Assert=] that |window| is an [=environment settings object=] whose [=global object=] is a {{Window}}.
+    1. Let |sharedStorage| be |window|'s [=global object=]'s {{Window/sharedStorage}}.
+    1. If |sharedStorage| is null, then return.
+    1. Let |list| be |response|'s [=response/header list=].
+    1. Let |operationsToParse| be the result of running [=get a structured field value=] algorithm given "[=Shared-Storage-Write=]", "`list`", and |list| as input.
+    1. If |operationsToParse| is null or [=list/empty=], then return.
+    1. For each tuple (|item|, |parameters|) in |operationsToParse|, perform the following steps:
+        1. If |item| is an [=Inner List=], continue.
+        1. [=Assert=] that |item| is an [=Bare Item=].
+        1. Let |operationString| be the result of running [=get the string value=] for |item|.
+        1. If |operationString| is failure, continue.
+        1. Switch on |operationString|:
+            <dl class=switch>
+              <dt> If |operationString| is "`clear`":
+                 <dd> Perform the following steps:
+                     1. Run {{WindowSharedStorage/clear()}} on |sharedStorage|.
```

Comment:
It's generally a bad idea to call WebIDL operations from inside an algorithm, since it leads to ambiguity if JavaScript replaces those functions. ... But it looks like this isn't actually a WebIDL operation even though it's defined `for` a WebIDL interface. In that case, it's probably more clear to just define it as an algorithm that doesn't look like a function call.

Reply:
Replaced